src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -72,37 +75,36 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
     try!(links::validate(deps));
 
     let dest = if build_config.release {"release"} else {"debug"};
-    let root = if resolve.root() == pkg.package_id() {
-        pkg
-    } else {
-        deps.iter().find(|p| p.package_id() == resolve.root()).unwrap()
-    };
+    let root = deps.iter().find(|p| p.package_id() == resolve.root()).unwrap();
```

Comment:
At the moment `-p SPEC` only builds packages that are dependencies of the "root" package, and all generated binaries are stored relative to that root package. I think this root package is mostly used to get a base path for the build output and I think this notation should still be valid?